`getByName("foo")` no longer throws `UnknownHostException`

### 0.4.7

```scala
//> using platform "native"
//> using nativeVersion "0.4.7"

import java.net.InetAddress

@main def main = InetAddress.getByName("foo")
```
```
java.net.UnknownHostException: foo: Name or service not known
        at java.lang.StackTrace$.currentStackTrace(Unknown Source)
        at java.lang.Throwable.fillInStackTrace(Unknown Source)
        at java.net.InetAddressBase.$anonfun$1(Unknown Source)
        at java.net.InetAddressBase$$Lambda$8.apply(Unknown Source)
        at scala.Option.getOrElse(Unknown Source)
        at java.net.InetAddressBase.getByName(Unknown Source)
        at java.net.InetAddress.getByName(Unknown Source)
        at test$package$.main(Unknown Source)
        at main.main(Unknown Source)
        at <none>.main(Unknown Source)
        at <none>.__libc_start_main(Unknown Source)
        at <none>._start(Unknown Source)
```

### 0.4.8

```
java.io.IOException: No address associated with hostname
        at java.lang.StackTrace$.currentStackTrace(Unknown Source)
        at java.lang.Throwable.fillInStackTrace(Unknown Source)
        at java.io.IOException.<init>(Unknown Source)
        at java.net.InetAddress$.getByNonNumericName$$anonfun$1(Unknown Source)
        at java.net.InetAddress$$$Lambda$6.apply(Unknown Source)
        at scala.scalanative.unsafe.Zone$.apply(Unknown Source)
        at java.net.InetAddress$.getByNonNumericName(Unknown Source)
        at java.net.InetAddress$.getByName$$anonfun$1(Unknown Source)
        at java.net.InetAddress$$$Lambda$8.apply(Unknown Source)
        at scala.Option.getOrElse(Unknown Source)
        at java.net.InetAddress$.getByName(Unknown Source)
        at java.net.InetAddress.getByName(Unknown Source)
        at test$package$.main(Unknown Source)
        at main.main(Unknown Source)
        at <none>.main(Unknown Source)
        at <none>.__libc_start_main(Unknown Source)
        at <none>._start(Unknown Source)
```

```
$ uname -a
Linux armanbilge-scalanative-7yqib28iyt2 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

I checkout https://github.com/scala-native/scala-native/commit/9b8f2cf6b1da1aa1c49c951a3a1f2474cf473e86 and run

```
sbt:scala-native> tests2_13/testOnly *.InetAddressTest
...
[error] Test javalib.net.InetAddressTest.getByName failed: java.lang.AssertionError: getByName(foo): unexpected exception type thrown; expected:<java.net.UnknownHostException> but was:<java.io.IOException>, took 0.153 sec
[error]     at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)
[error]     at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)
[error]     at scala.scalanative.unsafe.Zone$.apply(Unknown Source)
[error]     at java.lang.StackTrace$.currentStackTrace(Unknown Source)
[error]     at java.lang.Throwable.fillInStackTrace(Unknown Source)
[error]     at java.lang.AssertionError.this(Unknown Source)
[error]     ...
[error] Caused by: java.io.IOException: No address associated with hostname
[error]     at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)
[error]     at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)
[error]     at scala.scalanative.unsafe.Zone$.apply(Unknown Source)
[error]     at java.lang.StackTrace$.currentStackTrace(Unknown Source)
[error]     at java.lang.Throwable.fillInStackTrace(Unknown Source)
[error]     at java.io.IOException.this(Unknown Source)
[error]     at java.net.InetAddress$.$anonfun$getByNonNumericName$1(Unknown Source)
[error]     at java.net.InetAddress$$$Lambda$2.apply(Unknown Source)
[error]     at scala.scalanative.unsafe.Zone$.apply(Unknown Source)
[error]     at java.net.InetAddress$.getByNonNumericName(Unknown Source)
[error]     at java.net.InetAddress$.$anonfun$getByName$1(Unknown Source)
[error]     at java.net.InetAddress$$$Lambda$4.apply(Unknown Source)
[error]     at scala.Option.getOrElse(Unknown Source)
[error]     at java.net.InetAddress$.getByName(Unknown Source)
[error]     at java.net.InetAddress.getByName(Unknown Source)
[error]     at javalib.net.InetAddressTest.$anonfun$getByName$2(Unknown Source)
[error]     at javalib.net.InetAddressTest$$Lambda$8.apply(Unknown Source)
[error]     at org.scalanative.testsuite.utils.AssertThrows$$anon$2.run(Unknown Source)
[error]     ... 44 more
...

I think the 0.4.8 changes are doing their job, as time consuming that is for us.

The CI stack trace gives "Caused by: java.io.IOException: Temporary failure in name resolution"
I believe that would translate to EAI_AGAIN.  

Now the question becomes what do we want to do about it.  Lack of evidence (EAI_AGAIN) is
not strictly evidence for the lack of a name (EAI_NONAME), but both mean that no host translation is
available.  In the former, it is UnknowableHostException, the latter is a solid UnknownHostException.
Unlike an I/O read() call mumble_AGAIN, we probably do not want to loop trying again.

Now that we seem to have identified the underlying condition, is there a better way to handle 
the these Exceptions?


I have to say it or it will stick with me all day: There is something about "foo". Would make
a good title for a movie.

Lee, thank as always for the investigation and detailed write-up :) much appreciated. If you believe it is worthwhile to make this change then I like your proposal. I am also open to conceding that not every exception can behave the same across Native and JVM, and this is just one of those edge cases.

> I do have some lingering disquiet in that the text "Caused by: java.io.IOException: Temporary failure in name resolution"
looks like EAI_AGAIN and not the (-5, EAI_NODATA).

This is a good point. Let me verify that indeed this is the error message for that number.

--- a/javalib/src/main/scala/java/net/InetAddress.scala
+++ b/javalib/src/main/scala/java/net/InetAddress.scala
@@ -29,6 +29,8 @@ import scala.scalanative.posix.sys.socket._
 import scala.scalanative.posix.time.{time_t, time, difftime}
 import scala.scalanative.posix.unistd
 
+import scala.scalanative.meta.LinktimeInfo.{isLinux, isMac}
+
 /* Design note:
  *    Much of java.net, both in JVM and Scala Native defines or assumes
  *    the ipAddress field to have either 4 or 16 bytes.
@@ -333,7 +335,8 @@ object InetAddress {
       val gaiStatus = getaddrinfo(toCString(host), null, hints, addrinfo)
 
       if (gaiStatus != 0) {
-        if (gaiStatus == EAI_NONAME) {
+        val mappedStatus = mapGaiStatus(gaiStatus)
+        if (mappedStatus == EAI_NONAME) {
           val ifIndex = host.indexOf('%')
           val hasInterface = (ifIndex >= 0)
           if (!hasInterface) {
@@ -355,8 +358,8 @@ object InetAddress {
             )
           }
         } else {
-          val gaiMsg = SocketHelpers.getGaiErrorMessage(gaiStatus)
-          throw new IOException(gaiMsg)
+          val gaiMsg = SocketHelpers.getGaiErrorMessage(mappedStatus)
+          throw new UnknownHostException(host + ": " + gaiMsg)
         }
       } else
         try {
@@ -436,12 +439,7 @@ object InetAddress {
 
       if (gaiStatus != 0) {
         val gaiMsg = SocketHelpers.getGaiErrorMessage(gaiStatus)
-        val ex =
-          if (gaiStatus == EAI_NONAME)
-            new UnknownHostException(host + ": " + gaiMsg)
-          else
-            new IOException(gaiMsg)
-        throw ex
+        throw new UnknownHostException(host + ": " + gaiMsg)
       } else
         try {
           val preferIPv6 = SocketHelpers.getPreferIPv6Addresses()
@@ -624,9 +622,10 @@ object InetAddress {
       val gaiStatus = getaddrinfo(toCString(host), null, hints, ret)
 
       if (gaiStatus != 0) {
-        if (gaiStatus != EAI_NONAME) {
-          val gaiMsg = SocketHelpers.getGaiErrorMessage(gaiStatus)
-          throw new IOException(gaiMsg)
+        val mappedStatus = mapGaiStatus(gaiStatus)
+        if (mappedStatus != EAI_NONAME) {
+          val gaiMsg = SocketHelpers.getGaiErrorMessage(mappedStatus)
+          throw new UnknownHostException(host + ": " + gaiMsg)
         }
       } else
         try {
@@ -643,6 +642,35 @@ object InetAddress {
     val ptrInt = pb.asInstanceOf[Ptr[Int]]
     val ptrLong = pb.asInstanceOf[Ptr[Long]]
     (ptrInt(2) == 0xffff0000) && (ptrLong(0) == 0x0L)
+  }
+
+  private def mapGaiStatus(gaiStatus: Int): Int = {
+    /* This is where some arcane Operating System specific behavior
+     * comes to puddle and pool. This method is not for small children
+     * or maintainers with good taste & practice.
+     *
+     * EAI_NODATA was removed from RFC3493 "Basic Socket Interface Extensions
+     * for IPv6" in February 2003. EAI_NONAME was introduced and is the
+     * contemporary idiom. Although it is remove (i.e. well past deprecated),
+     * EAI_NODATA can be returned by Linux & macOS in some poorly defined
+     * circumstances.
+     *
+     * The magic integer values for Linux & macOS are hardcoded
+     * because they are extremely unlikely to change after all this time.
+     *
+     * For consistency of the reported message, map EAI_NODATA to EAI_NONAME.
+     * Both will return "UnknownHostException".
+     */
+
+    // EAI_NODATA was removed from FreeBSD a decade or more ago.
+    val EAI_NODATA =
+      if (isLinux) -5
+      else if (isMac) 7
+      else Integer.MAX_VALUE // placeholder, will never match
+
+    if (gaiStatus == EAI_NONAME) gaiStatus
+    else if (gaiStatus == EAI_NODATA) EAI_NONAME
+    else gaiStatus
   }
 
   def getAllByName(host: String): Array[InetAddress] = {

